cmd/root.go

```diff
@@ -2,56 +2,190 @@ package cmd
 
 import (
 	"fmt"
+	"image"
+	"io/fs"
 	"os"
+	"path/filepath"
+	"strings"
+	"sync"
 	"time"
 
 	"github.com/briandowns/spinner"
+	"github.com/gosuri/uilive"
 	"github.com/spf13/cobra"
 
 	"github.com/blackironj/panorama/conv"
 )
 
-const defaultEdgeLen = 1024
+const (
+	defaultEdgeLen      = 1024
+	maxConcurrentFiles  = 10 // Adjust this number based on your system's file descriptor limit
+)
 
 var (
 	inFilePath string
 	outFileDir string
+	inDirPath  string
 	edgeLen    int
+	sides      []string
 
+	validSides = []string{"front", "back", "left", "right", "top", "bottom"}
+	semaphore  = make(chan struct{}, maxConcurrentFiles)
+	progress   = struct {
+		sync.Mutex
+		totalFiles      int
+		processedFiles  int
+		startTime       time.Time
+		errors          []string
+	}{}
 	rootCmd = &cobra.Command{
 		Use:   "panorama",
 		Short: "convert equirectangular panorama img to Cubemap img",
 		Run: func(cmd *cobra.Command, args []string) {
-			if inFilePath == "" {
-				er("Need a image for converting")
-			}
-
-			fmt.Println("Read a image...")
-			inImage, ext, err := conv.ReadImage(inFilePath)
-			if err != nil {
-				er(err)
+			if inFilePath == "" && inDirPath == "" {
+				er("Need an input image file path or input directory")
 			}
 
-			s := spinner.New(spinner.CharSets[33], 100*time.Millisecond)
-			s.FinalMSG = "Complete converting!\n"
-			s.Prefix = "Converting..."
+			progress.startTime = time.Now()
 
-			s.Start()
-			canvases := conv.ConvertEquirectangularToCubeMap(edgeLen, inImage)
-			s.Stop()
+			if inFilePath != "" {
+				progress.totalFiles = 1
+				processSingleImage(inFilePath, outFileDir)
+			}
 
-			fmt.Println("Write images...")
-			if err := conv.WriteImage(canvases, outFileDir, ext); err != nil {
-				er(err)
+			if inDirPath != "" {
+				processDirectory(inDirPath, outFileDir)
```

Comment:
you wrote the following in the flag desc in the Readme.
```
  -i, --in string    input image file path (required if --indir is not specified)
  -d, --indir string input directory path (required if --in is not specified)
```
If this is not what you intended, how about modifying codes like this?
```go
if inFilePath != "" {
     progress.totalFiles = 1
     progressSingleImage(inFilePath, outFileDir)
}  else {
    processDirectory(inDirPath, outFileDir)
}
```

Reply:
```go
if len(inFilePath) > 0 && len(inDirPath) > 0 {
   er("Need only one path, not both")
}
```
I think that  codes what you added above are better. could you add that?